// @mui/x-data-grid/legacy/joy/joySlots.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
var _excluded = ["touchRippleRef", "inputProps", "onChange", "color", "size", "checked", "sx", "value", "inputRef"],
  _excluded2 = ["onChange", "label", "placeholder", "value", "inputRef", "type", "size", "variant"],
  _excluded3 = ["startIcon", "color", "endIcon", "size", "sx", "variant"],
  _excluded4 = ["color", "size", "sx", "touchRippleRef"],
  _excluded5 = ["name", "checkedIcon", "color", "disableRipple", "disableFocusRipple", "disableTouchRipple", "edge", "icon", "inputProps", "inputRef", "size", "sx", "onChange", "onClick"],
  _excluded6 = ["open", "onOpen", "value", "onChange", "size", "color", "variant", "inputProps", "MenuProps", "inputRef", "error", "native", "fullWidth", "labelId"],
  _excluded7 = ["native"],
  _excluded8 = ["shrink", "variant", "sx"];
import * as React from 'react';
import JoyCheckbox from '@mui/joy/Checkbox';
import JoyInput from '@mui/joy/Input';
import JoyFormControl from '@mui/joy/FormControl';
import JoyFormLabel from '@mui/joy/FormLabel';
import JoyButton from '@mui/joy/Button';
import JoyIconButton from '@mui/joy/IconButton';
import JoySwitch from '@mui/joy/Switch';
import JoySelect from '@mui/joy/Select';
import JoyOption from '@mui/joy/Option';
import { unstable_useForkRef as useForkRef } from '@mui/utils';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
function convertColor(color) {
  if (color === 'secondary') {
    return 'primary';
  }
  if (color === 'error') {
    return 'danger';
  }
  if (color === 'default' || color === 'inherit') {
    return 'neutral';
  }
  return color;
}
function convertSize(size) {
  return size ? {
    small: 'sm',
    medium: 'md',
    large: 'lg'
  }[size] : size;
}
function convertVariant(variant) {
  var defaultVariant = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'plain';
  if (!variant) {
    return defaultVariant;
  }
  return {
    standard: 'outlined',
    outlined: 'outlined',
    contained: 'solid',
    text: 'plain',
    filled: 'soft'
  }[variant] || defaultVariant;
}
var Checkbox = /*#__PURE__*/React.forwardRef(function (_ref, ref) {
  var touchRippleRef = _ref.touchRippleRef,
    inputProps = _ref.inputProps,
    onChange = _ref.onChange,
    color = _ref.color,
    size = _ref.size,
    checked = _ref.checked,
    sx = _ref.sx,
    value = _ref.value,
    inputRef = _ref.inputRef,
    props = _objectWithoutProperties(_ref, _excluded);
  return /*#__PURE__*/_jsx(JoyCheckbox, _extends({}, props, {
    slotProps: {
      input: _extends({}, inputProps, {
        ref: inputRef
      })
    },
    ref: ref,
    checked: checked,
    onChange: onChange
  }));
});
var TextField = /*#__PURE__*/React.forwardRef(function (_ref2, ref) {
  var _props$InputProps, _props$inputProps;
  var onChange = _ref2.onChange,
    label = _ref2.label,
    placeholder = _ref2.placeholder,
    value = _ref2.value,
    inputRef = _ref2.inputRef,
    type = _ref2.type,
    size = _ref2.size,
    variant = _ref2.variant,
    props = _objectWithoutProperties(_ref2, _excluded2);
  var rootRef = useForkRef(ref, (_props$InputProps = props.InputProps) == null ? void 0 : _props$InputProps.ref);
  var inputForkRef = useForkRef(inputRef, props == null ? void 0 : (_props$inputProps = props.inputProps) == null ? void 0 : _props$inputProps.ref);
  var _ref3 = props.InputProps || {},
    startAdornment = _ref3.startAdornment,
    endAdornment = _ref3.endAdornment;
  return /*#__PURE__*/_jsxs(JoyFormControl, {
    ref: rootRef,
    children: [/*#__PURE__*/_jsx(JoyFormLabel, {
      children: label
    }), /*#__PURE__*/_jsx(JoyInput, {
      type: type,
      value: value,
      onChange: onChange,
      placeholder: placeholder,
      variant: convertVariant(variant, 'outlined'),
      size: convertSize(size),
      slotProps: {
        input: _extends({}, props == null ? void 0 : props.inputProps, {
          ref: inputForkRef
        })
      },
      startDecorator: startAdornment,
      endDecorator: endAdornment
    })]
  });
});
var Button = /*#__PURE__*/React.forwardRef(function Button(_ref4, ref) {
  var startIcon = _ref4.startIcon,
    color = _ref4.color,
    endIcon = _ref4.endIcon,
    size = _ref4.size,
    sx = _ref4.sx,
    variant = _ref4.variant,
    props = _objectWithoutProperties(_ref4, _excluded3);
  return /*#__PURE__*/_jsx(JoyButton, _extends({}, props, {
    size: convertSize(size),
    color: convertColor(color),
    variant: convertVariant(variant),
    ref: ref,
    startDecorator: startIcon,
    endDecorator: endIcon,
    sx: sx
  }));
});
var IconButton = /*#__PURE__*/React.forwardRef(function IconButton(_ref5, ref) {
  var _convertColor;
  var color = _ref5.color,
    size = _ref5.size,
    sx = _ref5.sx,
    touchRippleRef = _ref5.touchRippleRef,
    props = _objectWithoutProperties(_ref5, _excluded4);
  return /*#__PURE__*/_jsx(JoyIconButton, _extends({}, props, {
    size: convertSize(size),
    color: (_convertColor = convertColor(color)) != null ? _convertColor : 'neutral',
    variant: "plain",
    ref: ref,
    sx: sx
  }));
});
var Switch = /*#__PURE__*/React.forwardRef(function Switch(_ref6, ref) {
  var name = _ref6.name,
    checkedIcon = _ref6.checkedIcon,
    colorProp = _ref6.color,
    disableRipple = _ref6.disableRipple,
    disableFocusRipple = _ref6.disableFocusRipple,
    disableTouchRipple = _ref6.disableTouchRipple,
    edge = _ref6.edge,
    icon = _ref6.icon,
    inputProps = _ref6.inputProps,
    inputRef = _ref6.inputRef,
    size = _ref6.size,
    sx = _ref6.sx,
    onChange = _ref6.onChange,
    onClick = _ref6.onClick,
    props = _objectWithoutProperties(_ref6, _excluded5);
  return /*#__PURE__*/_jsx(JoySwitch, _extends({}, props, {
    onChange: onChange,
    size: convertSize(size),
    color: convertColor(colorProp),
    ref: ref,
    slotProps: {
      input: _extends({}, inputProps, {
        name: name,
        onClick: onClick,
        ref: inputRef
      }),
      thumb: {
        children: icon
      }
    },
    sx: [_extends({}, edge === 'start' && {
      ml: '-8px'
    }, edge === 'end' && {
      mr: '-8px'
    })].concat(_toConsumableArray(Array.isArray(sx) ? sx : [sx]))
  }));
});
var Select = /*#__PURE__*/React.forwardRef(function (_ref7, ref) {
  var open = _ref7.open,
    onOpen = _ref7.onOpen,
    value = _ref7.value,
    onChange = _ref7.onChange,
    size = _ref7.size,
    color = _ref7.color,
    variant = _ref7.variant,
    inputProps = _ref7.inputProps,
    MenuProps = _ref7.MenuProps,
    inputRef = _ref7.inputRef,
    error = _ref7.error,
    native = _ref7.native,
    fullWidth = _ref7.fullWidth,
    labelId = _ref7.labelId,
    props = _objectWithoutProperties(_ref7, _excluded6);
  var handleChange = function handleChange(event, newValue) {
    if (event && onChange) {
      // Same as in https://github.com/mui/material-ui/blob/e5558282a8f36856aef1299f3a36f3235e92e770/packages/mui-material/src/Select/SelectInput.js#L288-L300

      // Redefine target to allow name and value to be read.
      // This allows seamless integration with the most popular form libraries.
      // https://github.com/mui/material-ui/issues/13485#issuecomment-676048492
      // Clone the event to not override `target` of the original event.
      var nativeEvent = event.nativeEvent || event;
      // @ts-ignore The nativeEvent is function, not object
      var clonedEvent = new nativeEvent.constructor(nativeEvent.type, nativeEvent);
      Object.defineProperty(clonedEvent, 'target', {
        writable: true,
        value: {
          value: newValue,
          name: props.name
        }
      });
      onChange(clonedEvent, null);
    }
  };
  return /*#__PURE__*/_jsx(JoySelect, _extends({}, props, {
    listboxOpen: open,
    onListboxOpenChange: function onListboxOpenChange(isOpen) {
      if (isOpen) {
        onOpen == null ? void 0 : onOpen({});
      } else {
        var _MenuProps$onClose;
        MenuProps == null ? void 0 : (_MenuProps$onClose = MenuProps.onClose) == null ? void 0 : _MenuProps$onClose.call(MenuProps, {}, undefined);
      }
    },
    size: convertSize(size),
    color: convertColor(color),
    variant: convertVariant(variant, 'outlined'),
    ref: ref,
    value: value,
    onChange: handleChange,
    slotProps: {
      button: {
        'aria-labelledby': labelId,
        ref: inputRef
      },
      listbox: {
        disablePortal: false,
        sx: {
          zIndex: 1350
        }
      }
    }
  }));
});
var Option = /*#__PURE__*/React.forwardRef(function (_ref8, ref) {
  var native = _ref8.native,
    props = _objectWithoutProperties(_ref8, _excluded7);
  return /*#__PURE__*/_jsx(JoyOption, _extends({}, props, {
    ref: ref
  }));
});
var InputLabel = /*#__PURE__*/React.forwardRef(function (_ref9, ref) {
  var shrink = _ref9.shrink,
    variant = _ref9.variant,
    sx = _ref9.sx,
    props = _objectWithoutProperties(_ref9, _excluded8);
  return /*#__PURE__*/_jsx(JoyFormLabel, _extends({}, props, {
    ref: ref,
    sx: sx
  }));
});
var joySlots = {
  baseCheckbox: Checkbox,
  baseTextField: TextField,
  baseButton: Button,
  baseIconButton: IconButton,
  baseSwitch: Switch,
  baseSelect: Select,
  baseSelectOption: Option,
  baseInputLabel: InputLabel,
  baseFormControl: JoyFormControl
  // BaseTooltip: MUITooltip,
  // BasePopper: MUIPopper,
};

export default joySlots;
